packages/toolbox-core/README.md

```diff
@@ -0,0 +1,440 @@
+> during initialization (e.g., `ToolboxClient(url, my_session)`).
+
+## Loading Tools
+
+You can load tools individually or in groups (toolsets) as defined in your
+Toolbox service configuration. Loading a toolset is convenient when working with
+multiple related functions, while loading a single tool offers more granular
+control.
+
+### Load a toolset
+
+A toolset is a collection of related tools. You can load all tools in a toolset
+or a specific one:
+
+```javascript
+// Load all tools
+const tools = await toolbox.loadToolset()
+
+// Load a specific toolset
+const tools = await toolbox.loadToolset("my-toolset")
+```
+
+### Load a single tool
+
+Loads a specific tool by its unique name. This provides fine-grained control.
+
+```javascript
+const tool = await toolbox.loadTool("my-tool")
+```
+
+## Invoking Tools
+
+Once loaded, tools behave like awaitable JS functions. You invoke them using
+`await` and pass arguments corresponding to the parameters defined in the tool's
+configuration within the Toolbox service.
+
+```javascript
+const tool = await toolbox.loadTool("my-tool")
+const result = await tool({a: 5, b: 2})
+```
+
+> [!TIP]
+> For a more comprehensive guide on setting up the Toolbox service itself, which
+> you'll need running to use this SDK, please refer to the [Toolbox Quickstart
+> Guide](https://googleapis.github.io/genai-toolbox/getting-started/local_quickstart).
+
+## Client to Server Authentication
+
+This section describes how to authenticate the ToolboxClient itself when
+connecting to a Toolbox server instance that requires authentication. This is
+crucial for securing your Toolbox server endpoint, especially when deployed on
+platforms like Cloud Run, GKE,  or any environment where unauthenticated access is restricted.
+
+This client-to-server authentication ensures that the Toolbox server can verify
+the identity of the client making the request before any tool is loaded or
+called. It is different from [Authenticating Tools](#authenticating-tools),
+which deals with providing credentials for specific tools within an already
+connected Toolbox session.
+
+### When is Client-to-Server Authentication Needed?
+
+You'll need this type of authentication if your Toolbox server is configured to
+deny unauthenticated requests. For example:
+
+- Your Toolbox server is deployed on Cloud Run and configured to "Require authentication."
+- Your server is behind an Identity-Aware Proxy (IAP) or a similar
+  authentication layer.
+- You have custom authentication middleware on your self-hosted Toolbox server.
+
+Without proper client authentication in these scenarios, attempts to connect or
+make calls (like `load_tool`) will likely fail with `Unauthorized` errors.
+
+### How it works
+
+The `ToolboxClient` allows you to specify functions (or coroutines for the async
+client) that dynamically generate HTTP headers for every request sent to the
+Toolbox server. The most common use case is to add an Authorization header with
```

Comment:
nit: it might be nice to link to https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Authorization